common/persistence/sql/sqlplugin/mysql/db.go

```diff
@@ -36,59 +39,88 @@ import (
 	mysqlschemaV8 "go.temporal.io/server/schema/mysql/v8"
 )
 
+// MySQL error codes
+const (
+	// ErrDupEntryCode MySQL Error 1062 indicates a duplicate primary key i.e. the row already exists,
+	// so we don't do the insert and return a ConditionalUpdate error.
+	ErrDupEntryCode = 1062
+
+	// Cannot execute statement in a READ ONLY transaction.
+	readOnlyTransactionCode = 1792
+	// Too many connections open
+	tooManyConnectionsCode = 1040
+	// Running in read-only mode
+	readOnlyModeCode = 1836
+)
+
 // db represents a logical connection to mysql database
 type db struct {
 	dbKind sqlplugin.DbKind
 	dbName string
 
-	db        *sqlx.DB
+	handle    *sqlplugin.DatabaseHandle
 	tx        *sqlx.Tx
-	conn      sqlplugin.Conn
 	converter DataConverter
 }
 
 var _ sqlplugin.AdminDB = (*db)(nil)
 var _ sqlplugin.DB = (*db)(nil)
 var _ sqlplugin.Tx = (*db)(nil)
 
-// ErrDupEntryCode MySQL Error 1062 indicates a duplicate primary key i.e. the row already exists,
-// so we don't do the insert and return a ConditionalUpdate error.
-const ErrDupEntryCode = 1062
+func isConnNeedsRefreshError(err error) bool {
+	myErr, ok := err.(*mysql.MySQLError)
+	if !ok {
+		return false
+	}
+	return myErr.Number == readOnlyModeCode || myErr.Number == readOnlyTransactionCode || myErr.Number == tooManyConnectionsCode
+}
 
 func (mdb *db) IsDupEntryError(err error) bool {
 	sqlErr, ok := err.(*mysql.MySQLError)
 	return ok && sqlErr.Number == ErrDupEntryCode
 }
 
+// Time to reset
+func (mdb *db) isConnNeedsRefreshError(err error) bool {
```

Comment:
what's the difference between this and the one on L70?

Reply:
Nothing that matters; it's leftover from an older version of the code